### ISSUE TITLE
Cycle detection, iterative version

### ISSUE THREAD

**Issue by AsileBcd** [JuliaGraphs/LightGraphs.jl#510]
Dear all,
Following PR #486 and #509 , I open a discussion about using a formal iterative procedure to build the algorithm of cycle detection.
I know it does not work, but my work is in attachment.
```julia
import Base: start, next, done

using LightGraphs

## New implementation.
type Circuits
    dg::DiGraph
end

type CircuitsState
    sccs::Vector{Vector{Int}}
    scc::Vector{Int}
    subdg::DiGraph
    vmap::Vector{Int}  
    startnode::Int
    blockedmap::Vector{Set{Int}}
    blocked::Vector{Bool}
    stack::Vector{Int}
    currentnode::Int
    done::Bool
end

"""
```unblock!(v::T, blocked::Vector{Bool}, B::Vector{Set{Int}})```

Unblock the vertices recursively.

# Arguments
* `v`: the vertex to unblock
* `blocked`: tell whether a vertex is blocked or not
* `B`: the map that tells if the unblocking of one vertex should unblock other vertices
"""
function unblock!(v::Int, blocked::Vector{Bool}, B::Vector{Set{Int}})
    blocked[v] = false
    for w in B[v]
        delete!(B[v], w)
        if blocked[w]
            unblock!(w, blocked, B)
        end
    end
end

function start(cr::Circuits) ## Problem if acyclic digraph
    sccs = strongly_connected_components(cr.dg)
    scc = pop!(sccs)
    while !isempty(sccs) && (length(scc) < 2)
        scc = pop!(sccs)
    end
    wdg, vmap = induced_subgraph(cr.dg, scc)
    b = falses(vertices(wdg)),
    bmap = [Set{Int}() for i in vertices(wdg)]
    neighbors = fadj(wdg, 1)
    state = CircuitsState(sccs, scc, wdg, vmap, 1, bmap, b, [1], 1, false)    
    state.blocked[1] = true
    return state
end

function done(cr::Circuits, state::CircuitsState)
    return isempty(state.sccs)
end

function next(cr::Circuits, state::CircuitsState)
    cycle = Vector{Int}()
    state.done = false
    push!(state.stack, state.currentnode)
    state.blocked[state.currentnode] = true
    for w in fadj(state.subdg, state.currentnode)
        if w == state.startnode
            state.done = true
            cycle = state.vmap[state.stack]
            unblock!(state.currentnode, state.blocked, state.blockedmap)
            pop!(state.stack)
        elseif !state.blocked[w]
            if !in(state.blockedmap[w], state.currentnode)
                push!(state.blockedmap[w], state.currentnode)
            end
            state.currentnode = w
            #next(cr, state)
        else
            if !in(state.blockedmap[w], state.currentnode)
                push!(state.blockedmap[w], state.currentnode)
            end
            pop!(state.stack)
            #next(cr, state)
        end
    end
    if state.done
        return cycle, state
        elseif (length(state.scc) > 1)
        state.currentnode = pop!(state.scc)
        #next(cr, state)
        elseif  !isempty(sccs)      
        state.scc = pop!(state.sccs)
        while length(state.scc) == 1
            state.scc = pop!(state.sccs)
        end
        state.subdg, state.vmap = induced_subgraph(state.dg, scc)
        #next(cr, state)
    end
    return cycle, state
end
```

**Comment by jpfairbanks** [comment]
Hi @AsileBcd,

Thanks for the work on this. I think it was a good idea to split these up. I am glad that we are taking the time to get this designed right. 

Do you have a specific question that you want answered?

When you say "Problem if acyclic digraph" what is the problem? Shouldn't the algorithm prove that there are no circuits and quit?

**Comment by AsileBcd** [comment]
Thank you!
For the previous part, I have no specific question, it really does not work anyway. I have tried since to work only on one strongly connected component, here is the code:
```julia
type SCCcircuits
    dg::DiGraph
end

type SCCcircuitState
    scc::Vector{Int}
    wdg::DiGraph
    vmap::Vector{Int}
    blocked::Vector{Bool}
    blockedmap::Vector{Set{Int}}
    startvertex::Int
    currentvertex::Int
    stack::Vector{Int}
    done::Bool
end

function done(cr::SCCcircuits, state::SCCcircuitState)
    length(state.scc) < 2
end

function start(cr::SCCcircuits)
    sccs = strongly_connected_components(cr.dg)
    scc = pop!(sccs)
    wdg, vmap = induced_subgraph(cr.dg, scc)
    b = falses(vertices(wdg))
    bmap = [Set{Int}() for i in vertices(wdg)]
    b[1] = true
    startvertex = 1
    currentvertex = 1
    stack = Int[]
    state = SCCcircuitState(scc, wdg, vmap, b, bmap, startvertex, currentvertex, stack, false)
    return state
end

function next(cr::SCCcircuits, state::SCCcircuitState)
    #cycle = Vector{Int}()
    state.done = false
    push!(state.stack, state.currentvertex)
    state.blocked[state.currentvertex] = true
    @show state.blocked, state.stack
    for neighbor in fadj(state.wdg, state.currentvertex)
        @show neighbor
        if neighbor == state.startvertex
            cycle = state.vmap[state.stack]
            @show cycle
            state.done = true
            break
            return cycle, state
            elseif !state.blocked[neighbor]
            state.currentvertex = neighbor
            next(cr, state)
        end
    end
    if state.done
        unblock!(state.currentvertex, state.blocked, state.blockedmap)
    else
        for neighbor in fadj(state.wdg, state.currentvertex)
            if !in(state.blockedmap(neighbor), state.currentvertex)
                push!(state.blockedmap(neighbor), state.currentvertex)
            end
        end
    end
    pop!(state.stack)
    return cycle, state
end
```
I have one problem (bug?), the `return cycle, state` after the `if` does not seem to work (the cycle is not outputted, whereas the `println(cycle)` prints the cycle. Do you have any idea why?
Thank you!

**Comment by jpfairbanks** [comment]
```julia
 if neighbor == state.startvertex
            cycle = state.vmap[state.stack]
            @show cycle
            state.done = true
            break
            return cycle, state
```
This part here? I think you want to remove the break keyword which terminates the loop immediately.

**Comment by AsileBcd** [comment]
Thanks, I should have been more careful when copy/past, it does not work when I remove the `break` too.

**Comment by jpfairbanks** [comment]
After a bit of modification, I got something reasonable. Here https://gist.github.com/jpfairbanks/9e2c104a87310f56959a0d47da07c811 is a gist with the code and output.

**Comment by AsileBcd** [comment]
I made some changes that avoid the problem that I had before, but I clearly have a problem once a cycle have been found to find the next one. Here is my new code (I have added a new element to the state).

```julia
function unblock!(v::Int, blocked::Vector{Bool}, B::Vector{Set{Int}})
    println("Entering unblocking $v")
    blocked[v] = false
    for w in B[v]
        delete!(B[v], w)
        if blocked[w]
            blocked, B = unblock!(w, blocked, B)
        end
    end
    @show blocked
    return blocked, B
end

type SCCcircuits
    dg::DiGraph
end

type SCCcircuitState
    scc::Vector{Int}
    wdg::DiGraph
    vmap::Vector{Int}
    blocked::Vector{Bool}
    blockedmap::Vector{Set{Int}}
    startvertex::Int
    currentvertex::Int
    stack::Vector{Int}
    done::Bool
    neighbors::Vector{Int}
end

function done(cr::SCCcircuits, state::SCCcircuitState)
    length(state.scc) < 2
end

function start(cr::SCCcircuits)
    sccs = strongly_connected_components(cr.dg)
    scc = pop!(sccs)
    wdg, vmap = induced_subgraph(cr.dg, scc)
    b = falses(vertices(wdg))
    bmap = [Set{Int}() for i in vertices(wdg)]
    b[1] = true
    startvertex = 1
    currentvertex = 1
    stack = [1]
    neighbors = fadj(wdg, currentvertex)
    state = SCCcircuitState(scc, wdg, vmap, b, bmap, startvertex, currentvertex, stack, false, neighbors)
    return state
end

function next(cr::SCCcircuits, state::SCCcircuitState)
    cycle = Vector{Int}()
    state.done = false
    #state.neighbors = fadj(state.wdg, state.currentvertex)
    @show state.blocked, state.stack
    while !isempty(state.neighbors)
        @show neighbor = pop!(state.neighbors)
        if neighbor == state.startvertex
            cycle = state.vmap[state.stack]
            println("found a cycle: $cycle")
            state.done = true
            @show state
            return cycle, state
            elseif !state.blocked[neighbor]
            state.currentvertex = neighbor
            state.blocked[state.currentvertex] = true
            push!(state.stack, state.currentvertex)
            state.neighbors = fadj(state.wdg, state.currentvertex)
            cycle, state = next(cr, state)
            break
            elseif isempty(state.neighbors)
            pop!(state.stack)
        end
    end
    if state.done
        @show state
        state.blocked, state.blockedmap = unblock!(state.currentvertex, state.blocked, state.blockedmap)
    else
        for neighbor in fadj(state.wdg, state.currentvertex)
            if !in(state.blockedmap(neighbor), state.currentvertex)
                push!(state.blockedmap(neighbor), state.currentvertex)
            end
        end
    end
    return cycle, state
end
```

**Comment by sbromberger** [comment]
Hi all, where do we stand on this?

**Comment by sbromberger** [comment]
Hi all, closing this out given our work in #717.

**Comment by AsileBcd** [comment]
Sorry to have not answered before. I was in holidays. No progresses were made anyway.
